internal/mtls/ca.go

```diff
@@ -0,0 +1,193 @@
+// future Cert providers, like:
+// - Vault
+// - Acme protocol
+// Keeping as an interface, so in future users can decice.
+type CAProvider interface {
+	GetName() string
+	GetCACertificate() (*CertificateGroup, error)
+	CreateRegistrationCertificate(name string) (map[string][]byte, error)
+}
+
+type TLSConfig struct {
+	config           *tls.Config
+	client           client.Client
+	caProvider       []CAProvider
+	Domains          []string
+	LocalhostEnabled bool
+	namespace        string
+}
+
+func NewMTLSconfig(client client.Client, namespace string, domains []string, localhostEnabled bool) *TLSConfig {
+	config := &TLSConfig{
+		config:           nil,
+		client:           client,
+		Domains:          domains,
+		namespace:        namespace,
+		LocalhostEnabled: localhostEnabled,
+	}
+
+	// Secret providers here
+	secretProvider := NewCASecretProvider(client, namespace)
+	config.caProvider = append(config.caProvider, secretProvider)
+
+	return config
+}
+
+// @TODO mainly used for testing, maybe not needed at all
+func (conf *TLSConfig) SetCAProvider(caProviders []CAProvider) {
+	conf.caProvider = caProviders
+}
+
+func (conf *TLSConfig) InitCertificates() (*tls.Config, []*x509.Certificate, error) {
+	if len(conf.caProvider) == 0 {
+		return nil, nil, fmt.Errorf("No provider set")
+	}
+
+	var errors error
+	caCerts := []*CertificateGroup{}
+
+	CACertChain := []*x509.Certificate{}
+	caCertPool := x509.NewCertPool()
+
+	for _, caProvider := range conf.caProvider {
+		caCert, err := caProvider.GetCACertificate()
+		if err != nil {
+			errors = multierror.Append(errors, fmt.Errorf(
+				"cannot get CA certificate for provider %s: %v",
+				caProvider.GetName(), err))
+			continue
+		}
+
+		caCerts = append(caCerts, caCert)
+		CACertChain = append(CACertChain, caCert.GetCert())
+		caCertPool.AppendCertsFromPEM(caCert.certPEM.Bytes())
+	}
+
+	if errors != nil {
+		return nil, nil, errors
+	}
+
+	if len(caCerts) == 0 {
+		return nil, nil, fmt.Errorf("Cannot get any CA certificate")
+	}
+
+	// We always sign the certificates with the first CA server. I guess that it's normal
+	serverCert, err := getServerCertificate(conf.Domains, conf.LocalhostEnabled, caCerts[0])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certificate, err := serverCert.GetCertificate()
+	if err != nil {
+		return nil, nil, fmt.Errorf("Cannot create server certfificate: %v", err)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{certificate},
+		ClientCAs:    caCertPool,
+		ClientAuth:   tls.RequireAnyClientCert,
+		MinVersion:   tls.VersionTLS13,
+	}
+	return tlsConfig, CACertChain, nil
+}
+
+func (conf *TLSConfig) CreateRegistrationClientCerts() error {
+
+	if len(conf.caProvider) == 0 {
+		return fmt.Errorf("Cannot get ca provider")
+	}
+
+	name := fmt.Sprintf("%s-%s",
+		regClientSecretNamePrefix,
+		utilrand.String(regClientSecretNameRandomLen))
+
+	certData, err := conf.caProvider[0].CreateRegistrationCertificate(name)
```

Comment:
Can you explain how will this behave with multiple CA providers since in `InitCertificates` we iterates over all of them and here only the first one is picked.

Reply:
you cannot sign a certificate with multiple CA.